### PR TITLE
Implement transactional session as part of the persistence

### PR DESCRIPTION
**Pull request by lailabougria**

### CODE REVIEW

File: src/NServiceBus.Persistence.Sql.TransactionalSession/SqlPersistenceOpenSessionOptions.cs [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    {
+        /// <summary>
+        ///
+        /// </summary>
+        /// <param name="tenantIdHeaderName"></param>
+        /// <param name="tenantId"></param>
+        public SqlPersistenceOpenSessionOptions(string tenantIdHeaderName = null, string tenantId = null)
+        {
+            var headers = new Dictionary<string, string>();
+            if (tenantIdHeaderName != null && tenantId != null)
```
Comment (SzymonPobiega) [inline src/NServiceBus.Persistence.Sql.TransactionalSession/SqlPersistenceOpenSessionOptions.cs:20]: Shouldn't we throw if one is null but the other is not?

**Review by SzymonPobiega — APPROVED** [review]

File: src/TransactionalSession.MsSqlMicrosoftDataClient.AcceptanceTests/SetUpFixture.cs [hunk 2]
```diff
@@ -0,0 +1,7 @@
+// using NUnit.Framework;
```
Comment (danielmarbach) [inline src/TransactionalSession.MsSqlMicrosoftDataClient.AcceptanceTests/SetUpFixture.cs:1]: Should this be deleted?

File: src/TransactionalSession.MsSqlMicrosoftDataClient.AcceptanceTests/When_running_with_multitenancy.cs [hunk 3]
```diff
@@ -0,0 +1,117 @@
+﻿namespace NServiceBus.TransactionalSession.AcceptanceTests;
```
Comment (danielmarbach) [inline src/TransactionalSession.MsSqlMicrosoftDataClient.AcceptanceTests/When_running_with_multitenancy.cs:1]: Block namespace

File: src/TransactionalSession.MsSqlMicrosoftDataClient.AcceptanceTests/When_using_transactional_session.cs [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+
+            Assert.AreEqual(rowId, result);
+        }
+
+        [TestCase(true)]
+        [TestCase(false)]
+        public async Task Should_not_send_messages_if_session_is_not_committed(bool outboxEnabled)
+        {
+            Context result = await Scenario.Define<Context>()
+                                           .WithEndpoint<AnEndpoint>(s => s.When(async (statelessSession, ctx) =>
```
Comment (danielmarbach) [inline src/TransactionalSession.MsSqlMicrosoftDataClient.AcceptanceTests/When_using_transactional_session.cs:123]: Do we need this crazy indentation?

File: src/TransactionalSession.MsSqlMicrosoftDataClient.AcceptanceTests/When_using_transactional_session.cs [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+            Assert.True(result.CompleteMessageReceived);
+            Assert.False(result.MessageReceived);
+        }
+
+        [TestCase(true)]
+        [TestCase(false)]
+        public async Task Should_send_immediate_dispatch_messages_even_if_session_is_not_committed(bool outboxEnabled)
+        {
+            Context result = await Scenario.Define<Context>()
+                                           .WithEndpoint<AnEndpoint>(s => s.When(async (_, ctx) =>
```
Comment (danielmarbach) [inline src/TransactionalSession.MsSqlMicrosoftDataClient.AcceptanceTests/When_using_transactional_session.cs:149]: Do we need this crazy indentation?

File: src/TransactionalSession.MsSqlSystemDataClient.AcceptenceTests/Infrastructure/TransactionSessionDefaultServerSystemData.cs [hunk 6]
```diff
@@ -0,0 +1,7 @@
+﻿namespace NServiceBus.TransactionalSession.AcceptanceTests;
```
Comment (danielmarbach) [inline src/TransactionalSession.MsSqlSystemDataClient.AcceptenceTests/Infrastructure/TransactionSessionDefaultServerSystemData.cs:1]: Block namespace

File: src/TransactionalSession.MsSqlSystemDataClient.AcceptenceTests/TransactionalSession.MsSqlSystemDataClient.AcceptenceTests.csproj [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    <PackageReference Include="Nunit" Version="3.13.3" />
+    <PackageReference Include="NServiceBus.AcceptanceTesting" Version="8.0.0-rc.1" />
+    <PackageReference Include="GitHubActionsTestLogger" Version="2.0.1" />
+    <PackageReference Include="NServiceBus.Persistence.Sql" Version="7.0.0-beta.2" />
+  </ItemGroup>
+
+  <Import Project="..\ScriptBuilderTask\NServiceBus.Persistence.Sql.targets" />
+
+  <ItemGroup>
+    <Compile Include="..\TransactionalSession.MsSqlMicrosoftDataClient.AcceptanceTests\Infrastructure\CaptureServiceProviderStartupTask.cs">
```
Comment (danielmarbach) [inline src/TransactionalSession.MsSqlSystemDataClient.AcceptenceTests/TransactionalSession.MsSqlSystemDataClient.AcceptenceTests.csproj:30]: Simplify to Infrastructure\*.cs ?

File: src/TransactionalSession.MsSqlSystemDataClient.AcceptenceTests/When_using_transactional_session.cs [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+
+            Assert.AreEqual(rowId, result);
+        }
+
+        [TestCase(true)]
+        [TestCase(false)]
+        public async Task Should_not_send_messages_if_session_is_not_committed(bool outboxEnabled)
+        {
+            Context result = await Scenario.Define<Context>()
+                                           .WithEndpoint<AnEndpoint>(s => s.When(async (statelessSession, ctx) =>
```
Comment (danielmarbach) [inline src/TransactionalSession.MsSqlSystemDataClient.AcceptenceTests/When_using_transactional_session.cs:123]: Do we need this crazy indentation?

File: src/TransactionalSession.MsSqlSystemDataClient.AcceptenceTests/When_using_transactional_session.cs [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+            Assert.True(result.CompleteMessageReceived);
+            Assert.False(result.MessageReceived);
+        }
+
+        [TestCase(true)]
+        [TestCase(false)]
+        public async Task Should_send_immediate_dispatch_messages_even_if_session_is_not_committed(bool outboxEnabled)
+        {
+            Context result = await Scenario.Define<Context>()
+                                           .WithEndpoint<AnEndpoint>(s => s.When(async (_, ctx) =>
```
Comment (danielmarbach) [inline src/TransactionalSession.MsSqlSystemDataClient.AcceptenceTests/When_using_transactional_session.cs:149]: Do we need this crazy indentation?

File: src/TestHelper/MsSqlSystemDataClientConnectionBuilder.cs [hunk 10]
```diff
@@ -124,12 +144,17 @@ static string GetConnectionString()
             throw new Exception("SQLServerConnectionString environment variable is empty");
         }
 
+        if (trustServerCertificate)
```
[the remark this answers is not in the document]
Comment (lailabougria) [reply src/TestHelper/MsSqlSystemDataClientConnectionBuilder.cs]: Removed this anyway for now in this PR.

File: src/TransactionalSession.MsSqlSystemDataClient.AcceptenceTests/TransactionalSession.MsSqlSystemDataClient.AcceptenceTests.csproj (hunk 7 above)
Comment (lailabougria) [reply src/TransactionalSession.MsSqlSystemDataClient.AcceptenceTests/TransactionalSession.MsSqlSystemDataClient.AcceptenceTests.csproj:30]: It's not all files I'm including though...

File: src/NServiceBus.Persistence.Sql.TransactionalSession/SqlPersistenceTransactionalSessionExtensions.cs [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+
+    /// <summary>
+    /// Enables the transactional session feature.
+    /// </summary>
+    public static class SqlPersistenceTransactionalSessionExtensions
+    {
+        /// <summary>
+        /// Enables transactional session for this endpoint.
+        /// </summary>
+        public static PersistenceExtensions<SqlPersistence> EnableTransactionalSession(
```
Comment (tmasternak) [inline src/NServiceBus.Persistence.Sql.TransactionalSession/SqlPersistenceTransactionalSessionExtensions.cs:14]: Should we add `Open` extension method to enable scenarios where no options are passed by the user?

Comment (danielmarbach) [reply src/NServiceBus.Persistence.Sql.TransactionalSession/SqlPersistenceTransactionalSessionExtensions.cs:14]: Same comment as with RavenDB.

File: src/NServiceBus.Persistence.Sql.TransactionalSession/NServiceBus.Persistence.Sql.TransactionalSession.csproj [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+      <AssemblyOriginatorKeyFile>$(SolutionDir)NServiceBus.snk</AssemblyOriginatorKeyFile>
+      <LangVersion>10.0</LangVersion>
+      <!-- We want the root namespace to match the transactional session one -->
+      <RootNamespace>NServiceBus.TransactionalSession</RootNamespace>
+      <Description>NServiceBus Transactional Session for SQL persistence.</Description>
+    </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Particular.Packaging" Version="2.2.0" PrivateAssets="All" />
+    <PackageReference Include="NServiceBus.TransactionalSession" Version="[2.0.0-beta.6, 3.0.0)" />
```
Comment (timbussmann) [inline src/NServiceBus.Persistence.Sql.TransactionalSession/NServiceBus.Persistence.Sql.TransactionalSession.csproj:15]:
```suggestion
    <PackageReference Include="NServiceBus.TransactionalSession" Version="[2.0.0-rc.1, 3.0.0)" />
```

File: src/NServiceBus.Persistence.Sql.TransactionalSession/SqlPersistenceOpenSessionOptions.cs [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+    {
+        /// <summary>
+        ///
+        /// </summary>
+        /// <param name="tenantIdHeaderName"></param>
+        /// <param name="tenantId"></param>
+        public SqlPersistenceOpenSessionOptions(string tenantIdHeaderName = null, string tenantId = null)
+        {
+            var headers = new Dictionary<string, string>();
+            if (tenantIdHeaderName != null)
```
Comment (timbussmann) [inline src/NServiceBus.Persistence.Sql.TransactionalSession/SqlPersistenceOpenSessionOptions.cs:20]: should this checks be using `IsNullOrEmpty` or `IsNullOrWhitespace`?

File: src/NServiceBus.Persistence.Sql.TransactionalSession/SqlPersistenceOpenSessionOptions.cs [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+            {
+                if (string.IsNullOrEmpty(tenantId))
+                {
+                    throw new Exception("A tenant header is available, but the value is missing.");
+                }
+                headers.Add(tenantIdHeaderName, tenantId);
```
Comment (timbussmann) [inline src/NServiceBus.Persistence.Sql.TransactionalSession/SqlPersistenceOpenSessionOptions.cs:26]: isn't setting headers the responsibility of the sender?

File: src/NServiceBus.Persistence.Sql.sln [hunk 15]
```diff
@@ -36,7 +36,8 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "PostgreSqlAcceptanceTests",
 EndProject
 Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "Solution Items", "Solution Items", "{FB898D09-ADF0-49D4-8089-D743F1A31D12}"
 	ProjectSection(SolutionItems) = preProject
-		Custom.Build.props = Custom.Build.props
+		Directory.Build.props = Directory.Build.props
+		..\GitVersion.yml = ..\GitVersion.yml
```
Comment (timbussmann) [inline src/NServiceBus.Persistence.Sql.sln:40]: why was this added here?

File: src/SqlPersistence/SqlPersistence.csproj [hunk 16]
```diff
@@ -17,6 +17,16 @@
     <PackageReference Include="Particular.Packaging" Version="2.2.0" PrivateAssets="All" />
   </ItemGroup>
 
+  <ItemGroup>
```
Comment (timbussmann) [inline src/SqlPersistence/SqlPersistence.csproj:20]: shouldn't there be a deleted `InternalsVisibleTo.cs` file in the PR changes?

File: src/TransactionalSession.MsSqlMicrosoftDataClient.AcceptanceTests/Infrastructure/TransactionSessionWithOutboxEndpoint.cs [hunk 17]
```diff
@@ -0,0 +1,20 @@
+﻿namespace NServiceBus.TransactionalSession.AcceptanceTests
+{
+    using System;
+    using System.Threading.Tasks;
+    using AcceptanceTesting.Support;
+
+    public class TransactionSessionWithOutboxEndpoint : TransactionSessionDefaultServer
+    {
+        public override Task<EndpointConfiguration> GetConfiguration(RunDescriptor runDescriptor, EndpointCustomizationConfiguration endpointConfiguration,
+                                                                     Func<EndpointConfiguration, Task> configurationBuilderCustomization) =>
```
Comment (timbussmann) [inline src/TransactionalSession.MsSqlMicrosoftDataClient.AcceptanceTests/Infrastructure/TransactionSessionWithOutboxEndpoint.cs:10]: that's some weird indentation

File: src/TransactionalSession.MsSqlSystemDataClient.AcceptenceTests/TransactionalSession.MsSqlSystemDataClient.AcceptenceTests.csproj (hunk 7 above)
Comment (timbussmann) [reply src/TransactionalSession.MsSqlSystemDataClient.AcceptenceTests/TransactionalSession.MsSqlSystemDataClient.AcceptenceTests.csproj:30]: > It's not all files I'm including though...

should the folder structure then be adjusted to allow easier sharing of common files?

File: src/TransactionalSession.MsSqlSystemDataClient.AcceptenceTests/When_using_transactional_session.cs [hunk 18]
```diff
@@ -0,0 +1,228 @@
...
+
+        [TestCase(true)]
+        [TestCase(false)]
+        public async Task Should_send_messages_and_insert_rows_in_synchronized_session_on_transactional_session_commit(
+            bool outboxEnabled)
+        {
+            string rowId = Guid.NewGuid().ToString();
+
+            await Scenario.Define<Context>()
+                          .WithEndpoint<AnEndpoint>(s => s.When(async (_, ctx) =>
```
Comment (timbussmann) [inline src/TransactionalSession.MsSqlSystemDataClient.AcceptenceTests/When_using_transactional_session.cs:29]: even here the identation should be fixed and aligned with other acceptance tests

File: src/TransactionalSession.Tests/TransactionalSession.Tests.csproj [hunk 19]
```diff
@@ -0,0 +1,24 @@
...
+  </PropertyGroup>
+
+
+  <ItemGroup>
+    <PackageReference Include="GitHubActionsTestLogger" Version="2.0.1" />
+    <PackageReference Include="NUnit" Version="3.13.3" />
+    <PackageReference Include="NUnit3TestAdapter" Version="4.2.1" />
+    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.3.1" />
+    <PackageReference Include="Particular.Approvals" Version="0.3.0" />
+    <PackageReference Include="PublicApiGenerator" Version="10.3.0" />
```
Comment (timbussmann) [inline src/TransactionalSession.Tests/TransactionalSession.Tests.csproj:16]: this should have an explicit (non-ranged) dependency on the tx session package so that dependabot will update tests to the latest verison

File: src/TransactionalSession.MsSqlMicrosoftDataClient.AcceptanceTests/TransactionalSession.MsSqlMicrosoftDataClient.AcceptanceTests.csproj [hunk 20]
```diff
@@ -0,0 +1,30 @@
...
+    <LangVersion>10.0</LangVersion>
+    <!-- We want the root namespace to match the transactional session one -->
+    <RootNamespace>NServiceBus.TransactionalSession.AcceptanceTests</RootNamespace>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <ProjectReference Include="..\ScriptBuilderTask\ScriptBuilderTask.csproj" />
+    <ProjectReference Include="..\SqlPersistence\SqlPersistence.csproj" />
+    <ProjectReference Include="..\TestHelper\TestHelper.csproj" />
+    <ProjectReference Include="..\NServiceBus.Persistence.Sql.TransactionalSession\NServiceBus.Persistence.Sql.TransactionalSession.csproj" />
```
Comment (timbussmann) [inline src/TransactionalSession.MsSqlMicrosoftDataClient.AcceptanceTests/TransactionalSession.MsSqlMicrosoftDataClient.AcceptanceTests.csproj:16]: this project should have an explicit non-ranged dependency on the tx session package so that dependabot will update it.

File: src/NServiceBus.Persistence.Sql.TransactionalSession/SqlPersistenceOpenSessionOptions.cs (hunk 13 above)
Comment (danielmarbach) [reply src/NServiceBus.Persistence.Sql.TransactionalSession/SqlPersistenceOpenSessionOptions.cs:20]: I think this should be a value tuple. You always need both values

File: src/NServiceBus.Persistence.Sql.sln (hunk 15 above)
Comment (danielmarbach) [reply src/NServiceBus.Persistence.Sql.sln:40]: Fixed it for Laila

File: src/SqlPersistence/SqlPersistence.csproj (hunk 16 above)
Comment (danielmarbach) [reply src/SqlPersistence/SqlPersistence.csproj:20]: Fixed it for Laila

File: src/TransactionalSession.MsSqlSystemDataClient.AcceptenceTests/When_using_transactional_session.cs (hunk 18 above)
Comment (danielmarbach) [reply src/TransactionalSession.MsSqlSystemDataClient.AcceptenceTests/When_using_transactional_session.cs]: I have done that already for Laila

File: src/TransactionalSession.MsSqlMicrosoftDataClient.AcceptanceTests/Infrastructure/TransactionSessionWithOutboxEndpoint.cs (hunk 17 above)
Comment (danielmarbach) [reply src/TransactionalSession.MsSqlMicrosoftDataClient.AcceptanceTests/Infrastructure/TransactionSessionWithOutboxEndpoint.cs:10]: Fixed it for Laila

File: src/TransactionalSession.MsSqlSystemDataClient.AcceptenceTests/TransactionalSession.MsSqlSystemDataClient.AcceptenceTests.csproj (hunk 7 above)
Comment (danielmarbach) [reply src/TransactionalSession.MsSqlSystemDataClient.AcceptenceTests/TransactionalSession.MsSqlSystemDataClient.AcceptenceTests.csproj]: I've done that already

File: src/NServiceBus.Persistence.Sql.TransactionalSession/SqlPersistenceOpenSessionOptions.cs (hunk 13 above)
Comment (danielmarbach) [reply src/NServiceBus.Persistence.Sql.TransactionalSession/SqlPersistenceOpenSessionOptions.cs]: Pushed a suggestion. Can always remove or tweak it again

**Review by danielmarbach — APPROVED** [review]

File: src/NServiceBus.Persistence.Sql.TransactionalSession/SqlPersistenceOpenSessionOptions.cs (hunk 14 above)
Comment (timbussmann) [reply src/NServiceBus.Persistence.Sql.TransactionalSession/SqlPersistenceOpenSessionOptions.cs]: > the connection to the right DB is established for the outbox records

that should be done via the control message and that's set on the metadata though?

Comment (danielmarbach) [reply src/NServiceBus.Persistence.Sql.TransactionalSession/SqlPersistenceOpenSessionOptions.cs]: The metadata is propagated to headers of the control message. We also need to add it to the headers of the IncomingMessage to make sure when the outbox interaction happens, the appropriate connection will be opened. Do you foresee a problem there? Otherwise, I'm not clear what you are hinting at. It might be though that I'm just not yet understanding where this is going.

Comment (timbussmann) [reply src/NServiceBus.Persistence.Sql.TransactionalSession/SqlPersistenceOpenSessionOptions.cs]: sorry I seem to have misread the headers part and the use of the extensions. Resolving this comment.

Comment (danielmarbach) [reply src/NServiceBus.Persistence.Sql.TransactionalSession/SqlPersistenceOpenSessionOptions.cs]: No worries. It is good to double and tripple check these things
